mozetl/engagement/retention/job.py

```diff
@@ -0,0 +1,166 @@
+    )
+    usage_hours = subsession_length / SECONDS_PER_HOUR
+
+    mapping = {
+        # attributes
+        'client_id': None,
+        'subsession_start': F.date_format(client_date, "yyyy-MM-dd"),
+        'profile_creation': F.date_format(pcd, "yyyy-MM-dd"),
+        'days_since_creation': (
+            F.when(days_since_creation < 0, F.lit(-1))  # -1 is the value for bad data
+            .otherwise(days_since_creation)
+            .cast("long")),
+        'channel': 'normalized_channel',
+        'app_version': None,
+        'geo': in_top_countries("country"),
+        'distribution_id': None,
+        'is_funnelcake': is_funnelcake,
+        'source': 'attribution.source',
+        'medium': 'attribution.medium',
+        'campaign': 'attribution.campaign',
+        'content': 'attribution.content',
+        'sync_usage': sync_usage(
+            "sync_count_desktop",
+            "sync_count_mobile",
+            "sync_configured"
+        ),
+        'is_active': client_date <= utils.to_datetime(F.lit(start_ds)),
+        # metrics
+        'usage_hours': usage_hours,
+        'sum_squared_usage_hours': usage_hours ** 2,
+        'total_uri_count':
+            F.col('scalar_parent_browser_engagement_total_uri_count').cast('long'),
+        'unique_domains_count':
+            F.col('scalar_parent_browser_engagement_unique_domains_count').cast('long'),
+    }
+    expr = utils.build_col_expr(mapping)
+
+    cleaned_data = (
+        main_summary
+        .select(expr)
+        .fillna({
+            'profile_creation': DEFAULT_DATE,
+            "days_since_creation": -1,
+        })
+        .fillna(0)
+        .fillna(0.0)
+        .fillna('unknown')
+        .select([field.name for field in retention_schema.fields])
+    )
+
+    return cleaned_data
+
+
+def save(cleaned, path):
+    (
+        cleaned
+        .write
+        .parquet(path, mode="overwrite")
+    )
+
+
+@click.command()
+@click.option('--start_date', required=True)
+@click.option('--path', default='retention_intermediate',
+              help="Path on hdfs to store intermediate data")
+@click.option('--input-bucket', default='telemetry-parquet',
+              help="input bucket containing main_summary")
+@click.option('--input-prefix', default='main_summary/v4',
+              help="input prefix containing main_summary")
+@click.option('--period', default=1,
+              help="length of the retention period in days")
+@click.option('--slack', default=2,
+              help="number of days to account for submission latency")
+@click.option('--sample/--no-sample', default=False)
+def main(start_date, path, input_bucket, input_prefix,
+         period, slack, sample):
+    """Compute churn / retention information for unique segments of
+    Firefox users acquired during a specific period of time.
+    """
+    spark = (
+        SparkSession
+        .builder
+        .appName("retention")
+        .getOrCreate()
+    )
```

Comment:
This formatting is a bit odd. I suggest to try to follow PEP8, e.g.:

```python
spark = SparkSession.builder.appName("retention").getOrCreate()
```

Reply:
I've seen this style used in Spark, [particularly in scala](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala#L39-L43). I guess it makes sense to keep things on a single line since we read left to right.